### PR TITLE
Snapshot API saves and restores time deltas.

### PR DESCRIPTION
**Pull request by scnale**
When running tests, the snapshot API wasn't saving the time adjustments made through the `evm_timeIncrease` JSON-RPC command. I added the code necessary to save and restore the appropriate state but I'm not sure this is the best way to do it. According to my own tests it seems to be working.

### CODE REVIEW

**Comment by hiqua** [comment]
Is there a reason not to merge these changes?

**Comment by madvas** [comment]
Please merge, much needed

**Comment by benjamincburns** [comment]
@scnale sorry it has taken us so long to circle 'round to this PR! Unfortunately, I don't think it's working. I've pushed a (failing) test to this PR which demonstrates how I think it's meant to work. Can you please either update the test to correct my misunderstanding, or update your solution to make the test pass?

**Comment by benjamincburns** [comment]
I don't know why, but the new test isn't showing in the latest travis build. I promise it's there, and if you run it locally it fails.

**Comment by scnale** [comment]
Ah, I see. Comparing timestamps won't work because the time adjustment is added to the current timestamp. I think the best way to test this would be to:
- take a snapshot
- call `evm_increaseTime` with any value
- restore the snapshot
- call `evm_increaseTime` with value `k` (preferably different from the previous one) and check its return value for equality with `k`.

**Comment by benjamincburns** [comment]
@scnale it's been quite a while - no pressure, but do you intend to fix up the test failures? If not, I'll take it on if I have time, or if I don't I'll just close this PR and flag the issue as `needshelp`.

**Comment by ouziel-slama** [comment]
Please fix/merge, much needed to write real tests. Thanks.

**Comment by benjamincburns** [comment]
@ouziel-slama, per comments above, the changes in this PR don't work as intended. Since I haven't heard back from @scnale in quite a while I'll close this to avoid confusion. If you'd like to fork @scnale's changes and submit a new PR yourself, I strongly encourage it!

**Comment by PhABC** [comment]
@benjamincburns and @scnale the fix works as intended. There is a mistake in the new test case, where revert should revert to snapshots[1]. Line `87` should therefore be `send("evm_revert",[1], function(err, result) { ... });` instead of `send("evm_revert", function(err, result) { ... });`. 

With this change, all the tests related to this PR pass. The original `evm_revert` call does properly revert to the very first `timeAdjustement`, as expected as well. Indeed, calling `evm_increaseTime` at line 46  
increments the `timeAdjustment` by `secondsToJump` seconds, a change that persists throughout the remaining tests. When calling `send("evm_revert", function(err, result) { ... });`, `timeAdjustment` is set to `originalTimeAdjustment - secondsToJump`, as expected. 

### Trace with current test implementation

1. `timeAdjustement` : -50652919 
2. `evm_increaseTime` (line 46)
3. `timeAdjustement` : -50634919
4. `evm_snapshot`: (line 75, snapshot 1, `timeAdjustement` == -50634502)
5. `evm_increaseTime` (line 77)
6. `timeAdjustement` : -50616919
7. `send("evm_revert", function(err, result) {...})` (line 87)
8. `timeAdjustement` : -50616919

The current test assert that `4.` should be equal to `8`, which is false. By specifying snapshot to 1, the timeAdjustment is correctly reverted to the value stored at `4.`

**Comment by benjamincburns** [comment]
> There is a mistake in the new test case

But... but I wrote that test case! :blush:

Thanks so much for pointing this out, @PhABC - I'll have a look!

File: test/time_adjust.js [hunk 1]
```diff
@@ -64,4 +64,34 @@ describe('Time adjustment', function() {
       })
     })
   })
+
+  it('should revert time adjustments when snapshot is reverted', function(done) {
+    // Adjust time
+    web3.eth.getBlock('latest', function(err, block){
+      if(err) return done(err)
+      var previousBlockTime = block.timestamp
```
Comment (graemecode) [inline test/time_adjust.js]: Unused var, fyi

**Pull request merged**